File: leetcode-sync/0167-two-sum-ii---input-array-is-sorted/solution.py
class Solution:
    def twoSum(self, numbers: list[int], target: int) -> list[int]:
        if target <= 0:
            for i, x in enumerate(numbers):
                for j, y in enumerate(numbers[i + 1:]):
                    if x + y < target:
                        break

                    if x + y == target:
                        return [i + 1, i + j + 2]

        numbers_len = len(numbers)

        if target == 0:
            # Could implement simeoultenaous search from start and end of the list
            for i, x in enumerate(numbers):
                for j, y in enumerate(numbers[i + 1:]):
                    if x + y == target:
                        return [i + 1, i + j + 2]
                    
        numbers.reverse()

        if target >= 0:
            for i, x in enumerate(numbers):
                if x > target:
                    continue

                for j, y in enumerate(numbers[i + 1:]):
                    if x + y == target:
                        return [numbers_len - (i + j + 1), numbers_len - i]

        

File: leetcode-sync/0167-two-sum-ii---input-array-is-sorted/solution.java
import java.util.*;

class Solution {
    public int[] v1(int[] numbers, int target) {
        // Naive approach, works the same for sorted and unsorted array.
        Map<Integer, Integer> numsIndexes = new HashMap<>(numbers.length);

        int i = 0;
        for (int num : numbers) {
            if (numsIndexes.keySet().contains(target - num)) {
                return new int[] { numsIndexes.get(target - num) + 1, i + 1 };
            }

            numsIndexes.put(num, i++);
        }

        return new int[] { -1, -1 };
    }

    public int[] v2(int[] numbers, int target) {
        int left = 0;
        int right = numbers.length - 1;

        while (left < right) {
            int currentSum = numbers[left] + numbers[right];

            if (currentSum < target) {
                left++;
                continue;
            } else if (currentSum > target){
                right--;
                continue;
            }

            return new int[] { left + 1, right + 1 };
        }

        return new int[] { -1, -1 };

    }

    public int[] twoSum(int[] numbers, int target) {
        return v2(numbers, target);
    }
}
